stmhal/rtc.c

```diff
@@ -52,119 +54,119 @@ static mp_uint_t rtc_info;
 
 // Note: LSI is around (32KHz), these dividers should work either way
 // ck_spre(1Hz) = RTCCLK(LSE) /(uwAsynchPrediv + 1)*(uwSynchPrediv + 1)
+// modify RTC_ASYNCH_PREDIV & RTC_SYNCH_PREDIV in board/<BN>/mpconfigport.h to change sub-second ticks
+// default is 3906.25 µs, min is ~30.52 µs (will increas Ivbat by ~500nA)
+#ifndef RTC_ASYNCH_PREDIV
 #define RTC_ASYNCH_PREDIV (0x7f)
+#endif
+#ifndef RTC_SYNCH_PREDIV
 #define RTC_SYNCH_PREDIV  (0x00ff)
+#endif
 
-#if 0
-#define RTC_INFO_USE_EXISTING (0)
-#define RTC_INFO_USE_LSE (1)
-#define RTC_INFO_USE_LSI (3)
-
-void rtc_init(void) {
-    // Enable the PWR clock
-    RCC_APB1PeriphClockCmd(RCC_APB1Periph_PWR, ENABLE);
-
-    // Allow access to RTC
-    PWR_BackupAccessCmd(ENABLE);
-
-    if (RTC_ReadBackupRegister(RTC_BKP_DR0) == 0x32F2) {
-        // RTC still alive, so don't re-init it
-        // wait for RTC APB register synchronisation
-        RTC_WaitForSynchro();
-        rtc_info = RTC_INFO_USE_EXISTING;
-        return;
+/*
+ * lse_magic
+ *
+ */
+bool lse_magic(void) {
+#ifdef DO_LSE_BLACK_MAGIC
+    uint32_t mode_in = GPIOC->MODER & 0x3fffffff;
+    uint32_t mode_out = mode_in | 0x40000000;
+    GPIOC->MODER = mode_out;
+    GPIOC->OTYPER &= 0x7fff;
+    GPIOC->BSRRH = 0x8000;
+    GPIOC->OSPEEDR &= 0x3fffffff;
+    GPIOC->PUPDR &= 0x3fffffff;
+    int i = 0xff0;
+    __IO int d = 0;
+    uint32_t tc = 0;
+    __IO uint32_t j;
+    while (i) {
+        GPIOC->MODER = mode_out;
+        GPIOC->MODER = mode_in;
+        for (j = 0; j < d; j++) ;
+        i--;
+        if ((GPIOC->IDR & 0x8000) == 0) {
+            tc++;
+        }
     }
+    return (tc < 0xff0)?true:false;
+#else
+    return false;
+#endif
+}
 
-    uint32_t timeout = 10000000;
-
-    // Enable the PWR clock
-    RCC_APB1PeriphClockCmd(RCC_APB1Periph_PWR, ENABLE);
+STATIC void RTC_CalendarConfig(void);
+STATIC void TSP_RTC_MspKick(RTC_HandleTypeDef *hrtc, bool rtc_use_lse);
```

Comment:
I would rename these functions to remove the TSP (assuming that's your initials).  Better to name it after its functionality, not the author :)